src/ast/passes/semantic_analyser.cpp

```diff
@@ -2285,21 +2327,18 @@ void SemanticAnalyser::visit(For &f)
   auto *mapkey = get_map_key_type(map);
   auto *mapval = get_map_type(map);
 
-  if (mapkey && mapkey->args_.size() == 0) {
-    LOG(ERROR, map.loc, err_)
-        << "Maps used as for-loop expressions must have keys to iterate over";
+  if (!mapkey || mapkey->IsNoneTy()) {
```

Comment:
```suggestion
  if (mapkey && mapkey->IsNoneTy()) {
```
This logic can be kept similar to the old code, right? (I might be missing something)

I try to return error messages as soon as possible and only resort to `is_final_pass` when absolutely necessary. That way the most useful error messages tend to get printed out first, and fewer bogus messages will be displayed to the user (i.e. those errors which were deferred to the final pass, if we error out before the final pass)